src/main/java/com/exasol/adapter/dynamodb/documentpath/DocumentPathWalker.java

```diff
@@ -0,0 +1,129 @@
+package com.exasol.adapter.dynamodb.documentpath;
+
+import java.util.ArrayList;
+import java.util.List;
+import java.util.function.Function;
+
+import com.exasol.adapter.dynamodb.documentnode.DocumentArray;
+import com.exasol.adapter.dynamodb.documentnode.DocumentNode;
+import com.exasol.adapter.dynamodb.documentnode.DocumentObject;
+
+/**
+ * This class walks a given path defined in {@link DocumentPathExpression} through a {@link DocumentNode} structure.
+ */
+@java.lang.SuppressWarnings("squid:S119") // VisitorType does not fit naming conventions.
+public class DocumentPathWalker<VisitorType> {
+    private final DocumentPathExpression pathExpression;
+
+    /**
+     * Creates an instance of {@link DocumentPathWalker}.
+     * 
+     * @param pathExpression path to walk
+     */
+    public DocumentPathWalker(final DocumentPathExpression pathExpression) {
+        this.pathExpression = pathExpression;
+    }
+
+    /**
+     * Walks the path defined in constructor through the given document.
+     * 
+     * @param rootNode document to walk through
+     * @return documents attributes described in {@link DocumentPathExpression}.
+     * @throws DocumentPathWalkerException if defined path does not exist in the given document
+     */
+    public List<DocumentNode<VisitorType>> walk(final DocumentNode<VisitorType> rootNode)
+            throws DocumentPathWalkerException {
+        return this.walk(rootNode, 0);
+    }
+
+    private List<DocumentNode<VisitorType>> walk(final DocumentNode<VisitorType> thisNode, final int position)
+            throws DocumentPathWalkerException {
+        if (this.pathExpression.size() <= position) {
+            return List.of(thisNode);
+        }
+        final Function<DocumentNode<VisitorType>, List<DocumentNode<VisitorType>>> converter = getConverterFor(
+                this.pathExpression.getPath().get(position));
+        try {
+            final List<DocumentNode<VisitorType>> nextNodes = converter.apply(thisNode);
+            final ArrayList<DocumentNode<VisitorType>> results = new ArrayList<>();
+            for (final DocumentNode<VisitorType> nextNode : nextNodes) {
```

Comment:
Since you do not use nextNodes outside of the for loop, you can apply converter directly?

```suggestion
            final ArrayList<DocumentNode<VisitorType>> results = new ArrayList<>();
            for (final DocumentNode<VisitorType> nextNode : converter.apply(thisNode)) {
```